web_poet/page_inputs.py

```diff
@@ -76,10 +99,88 @@ class ResponseData:
     ``status`` should represent the int status code of the HTTP response.
 
     ``headers`` should contain the HTTP response headers.
+
+    ``encoding`` encoding of the response. If None (default), encoding
+    is auto-detected from headers and body content.
     """
 
-    url: str
-    html: str
-    body: Optional[HttpResponseBody] = None
-    status: Optional[int] = None
-    headers: Optional[HttpResponseHeaders] = None
+    url: str = attrs.field()
+    # FIXME: raise an error or handle the case when str is passed as body
+    body: HttpResponseBody = attrs.field(converter=HttpResponseBody)
+    status: Optional[int] = attrs.field(default=None)
+    headers: Optional[HttpResponseHeaders] = attrs.field(default=None)
+    _encoding: Optional[str] = attrs.field(default=None)
+
+    _DEFAULT_ENCODING = 'ascii'
+    _cached_text: Optional[str] = None
+
+    @property
+    def text(self) -> str:
+        """
+        Content of the HTTP body, converted to unicode
+        using the detected encoding of the response, according
+        to the web browser rules (respecting Content-Type header, etc.)
+        """
+        # Access self.encoding before self._cached_text, because
+        # there is a chance self._cached_text would be already populated
+        # while detecting the encoding
+        encoding = self.encoding
+        if self._cached_text is None:
+            fake_content_type_header = f'charset={encoding}'
+            encoding, text = html_to_unicode(fake_content_type_header, self.body)
+            self._cached_text = text
+        return self._cached_text
+
+    @property
+    def encoding(self):
+        """ Encoding of the response """
+        return (
+            self._encoding
+            or self._headers_declared_encoding()
+            or self._body_declared_encoding()
+            or self._body_inferred_encoding()
+        )
+
+    @property
+    @memoizemethod_noargs
+    def selector(self) -> parsel.Selector:
+        # XXX: should we pass base_url=self.url, as Scrapy does?
+        return parsel.Selector(text=self.text)
+
+    def xpath(self, query, **kwargs):
+        return self.selector.xpath(query, **kwargs)
+
+    def css(self, query):
+        return self.selector.css(query)
```

Comment:
The `selector`, `xpath`, and `css` methods coincide with the `ResponseShortcutsMixin` like you've said. I don't believe they should be here as `ResponseData` is only concerned about representing the data compared to interacting with it. The `ResponseShortcutsMixin` should possess this ability.

Reply:
I think the difference is that with these methods being in the response, you don't need to wrap responses in page objects anymore. Example form https://github.com/scrapinghub/web-poet/pull/22:

```py
    # Simulates clicking on a button that says "View All Images"
    response: web_poet.ResponseData = await self.http_client.get(
        f"https://api.example.com/v2/images?id={item['product_id']}"
    )
    page = web_poet.WebPage(response)
    item["images"] = page.css(".product-images img::attr(src)").getall()
```

can be rewritten as

```py
    # Simulates clicking on a button that says "View All Images"
    response = await self.http_client.get(
        f"https://api.example.com/v2/images?id={item['product_id']}"
    )
    item["images"] = response.css(".product-images img::attr(src)").getall()
```

It seems putting these methods to response itself may simplify many similar examples. 

In general, separation of concerns is good, but I'm not sure what are we separating here, by leaving ResponseData as a simple data container without any logic. Page objects still contain website-specific code. HttpResponse is still not tied to any particular framework, and doesn't imply any particular download method. It also makes the API more familiar, as it's very close to Scrapy's. 

ResponseShortcutsMixin is used as a mixin to pages, e.g. WebPage uses it. It makes certain implicit assumptions: there is a "response" attribute, named "response" and containing ResponseData. It enables the usage like the following:

```py
class MyPage(WebPage):
    def to_item(self):
        price = self.css(".price").get()
```

With methods available on responses, we don't need to make these implicit assumptions:
```py
class MyPage(ItemPage):
    def to_item(self):
        price = self.response.css(".price").get()
```

We can still provide some shortcut to make methods available on Page level, but it becomes less important.

Reply:
Getting back to the example:

```py
@attrs.define
class MyPage(ItemPage, ReponseShortcutsMixin):
    response: HttpResponse

    def to_item(self):
        price = self.css(".price").get()
```

I really don't like how concerns are separated here. MyPage class defines "response" attribute, but this attribute is actually managed by ReponseShortcutsMixin. The code is not "local": to understand how it works, you need to figure out that self.css comes from ReponseShortcutsMixin, and that ReponseShortcutsMixin uses MyPage.response, there is a strong bi-directional connection between MyPage and ReponseShortcutsMixin. 

You don't have any of these issues with

```py
@attrs.define
class MyPage(ItemPage):
    response: HttpResponse

    def to_item(self):
        price = self.response.css(".price").get()
```

It's very top-down in this example: MyPage defines response, and can change it whatever needed - rename the variable, use 2 different responses (if supported), etc. to_item method uses response to extract data from it.

Reply:
> you don't need to wrap responses in page objects anymore

That's a good point. Not having to do `page = web_poet.WebPage(response)` everytime is very convenient! I also see your point about `ReponseShortcutsMixin`.

Thanks for eloquently explaining this. Moving these selector methods here makes sense. 👍 

What do you think about structuring it like this?

``` python
class HttpResponse:
    url
    body
    status
    headers

class ResponseData(HttpResponse):

    def selector(self):
        ...

    def xpath(self):
        ...

    def css(self):
        ...
```

In this way, we still have access to the bare minimum attributes of `HttpResponse` that could represent simple stuff like API calls returning plaintext which wouldn't necessarily need the selectors functionality in this case.

Moreover, `HEAD` requests also have their responses without a **body**, so representing them with methods like `selector` wouldn't make sense.

Nonetheless, we'd still keep `ResponseData` as the default recommended container. `HttpResponse` is also available on its own to be open to obscure use cases like the ones mentioned above.

Reply:
> What do you think about structuring it like this?

Not sure about this. What would be an advantage of having a bare data container, and a subclass with methods? User can always use a subclass with methods instead of a barebone class, so why a separate class is needed? 

I was thinking about a big backwards-incompatible change: removing ResponseData container completely. 

> Moreover, HEAD requests also have their responses without a body, so representing them with methods like selector wouldn't make sense.

Selectors won't work in this case, but they won't work in many other cases. For example, on json responses .json method would work, while .css won't. On text responses none of these methods would work, but response.text would be useful, as it'd decode the content properly. There can be GET requests which return a response without body (which is the same as empty body, right?).

Reply:
> There can be GET requests which return a response without body (which is the same as empty body, right?).

You have a point there.

> What would be an advantage of having a bare data container, and a subclass with methods? User can always use a subclass with methods instead of a barebone class, so why a separate class is needed?

Aside from semantically separating the HTTP attributes against its selector/extractor counterparts there's not much strong advantages for them.

I was initially thinking of potentially being able to leverage bare HTTP data containers on its own for some other obscure use cases. But after re-reading our discussion, I realized that majority of the use cases that web-poet need to support is about extracting data from responses as opposed to storing them. 

With that, I think it makes sense to store the HTTP attributes and its selectosr in a single class as currently proposed. 👍